File: test-platform-v2/backend/app/api/v2/ai_closed_loop.py
"""AITDE v2 AI QA Closed Loop API (V38).

Failure triage, healing apply, flaky, strategy performance, scenario gap,
suggestion inbox, human feedback and offline model evaluation. Mounted under
``/api/v2`` and feature-gated. Safety invariants are enforced in the service
layer — the router only passes through and never broadens them.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.v2.deps import require_aitde_v3
from app.core.deps import CurrentUser, get_db, require_permission
from app.modules.aitde.ai_closed_loop import service
from app.modules.aitde.ai_closed_loop.schemas import (
    FeedbackIn,
    GapConvertIn,
    HealingApplyIn,
    HypothesisReviewIn,
    ModelEvaluationIn,
    SuggestionReviewIn,
    TriageIn,
)
from app.schemas.common import R

router = APIRouter(
    tags=["AITDE - AI QA Closed Loop"], dependencies=[Depends(require_aitde_v3)]
)


def _issue_404(msg: str) -> None:
    from app.core.exceptions import APIException

    raise APIException(code=404, msg=msg, http_status=404)


@router.post("/runs/{run_id}/triage", response_model=R[dict])
def triage_run(
    run_id: int,
    payload: TriageIn,
    current: CurrentUser = Depends(require_permission("execution:detail")),
    db: Session = Depends(get_db),
):
    try:
        return R.ok(
            service.FailureTriageAgent.triage(
                db,
                run_id,
                payload.context,
                payload.model_ref,
                payload.prompt_version,
            )
        )
    except ValueError as exc:
        _issue_404(str(exc))


@router.get("/runs/{run_id}/hypotheses", response_model=R[list])
def list_hypotheses(
    run_id: int,
    current: CurrentUser = Depends(require_permission("execution:detail")),
    db: Session = Depends(get_db),
):
    return R.ok(service.FailureTriageAgent.list_hypotheses(db, run_id))


@router.post(
    "/hypotheses/{hypothesis_id}/review", response_model=R[dict]
)
def review_hypothesis(
    hypothesis_id: int,
    payload: HypothesisReviewIn,
    current: CurrentUser = Depends(require_permission("execution:update")),
    db: Session = Depends(get_db),
):
    try:
        return R.ok(
            service.HypothesisReviewService.review(
                db, hypothesis_id, payload.status, payload.reviewed_by, payload.reason
            )
        )
    except ValueError as exc:
        _issue_404(str(exc))


@router.post("/healing-proposals/{proposal_id}/apply", response_model=R[dict])
def apply_healing(
    proposal_id: int,
    payload: HealingApplyIn,
    current: CurrentUser = Depends(require_permission("execution:update")),
    db: Session = Depends(get_db),
):
    try:
        return R.ok(
            service.ApprovedHealingApply.apply(
                db, proposal_id, payload.approved_by, payload.note
            )
        )
    except ValueError as exc:
        _issue_404(str(exc))


@router.get("/flaky", response_model=R[list])
def list_flaky(
    scenario_adapter_id: int | None = Query(default=None),
    current: CurrentUser = Depends(require_permission("execution:list")),
    db: Session = Depends(get_db),
):
    return R.ok(service.FlakyClusterService.list(db, scenario_adapter_id))


@router.get("/scenarios/{scenario_id}/stability", response_model=R[dict])
def scenario_stability(
    scenario_id: int,
    current: CurrentUser = Depends(require_permission("mission:detail")),
    db: Session = Depends(get_db),
):
    return R.ok(service.FlakyClusterService.stability(db, scenario_id))


@router.get("/ai-suggestions", response_model=R[list])
def list_suggestions(
    status: str | None = Query(default=None),
    current: CurrentUser = Depends(require_permission("execution:list")),
    db: Session = Depends(get_db),
):
    return R.ok(
        service.SuggestionInboxService.list(db, current.project_id or 0, status)
    )


@router.post("/ai-suggestions/{suggestion_id}/review", response_model=R[dict])
def review_suggestion(
    suggestion_id: int,
    payload: SuggestionReviewIn,
    current: CurrentUser = Depends(require_permission("execution:update")),
    db: Session = Depends(get_db),
):
    try:
        return R.ok(
            service.SuggestionInboxService.review(
                db, suggestion_id, payload.status, payload.reviewed_by, payload.reason
            )
        )
    except ValueError as exc:
        _issue_404(str(exc))


@router.get("/missions/{mission_id}/scenario-gaps", response_model=R[list])
def list_gaps(
    mission_id: int,
    current: CurrentUser = Depends(require_permission("mission:detail")),
    db: Session = Depends(get_db),
):
    return R.ok(service.ScenarioGapDetector.list(db, mission_id))


@router.post("/scenario-gaps/{gap_id}/convert", response_model=R[dict])
def convert_gap(
    gap_id: int,
    payload: GapConvertIn,
    current: CurrentUser = Depends(require_permission("mission:update")),
    db: Session = Depends(get_db),
):
    try:
        return R.ok(
            service.ScenarioGapDetector.convert(
                db, gap_id, payload.title, payload.risk_level
            )
        )
    except ValueError as exc:
        _issue_404(str(exc))


@router.post("/feedback", response_model=R[dict])
def add_feedback(
    payload: FeedbackIn,
    current: CurrentUser = Depends(require_permission("execution:list")),
    db: Session = Depends(get_db),
):
    return R.ok(
        service.HumanFeedbackService.add(
            db, {**payload.model_dump(), "created_by": current.user.id}
        )
    )


@router.get("/feedback", response_model=R[list])
def list_feedback(
    target_type: str | None = Query(default=None),
    current: CurrentUser = Depends(require_permission("execution:list")),
    db: Session = Depends(get_db),
):
    return R.ok(
        service.HumanFeedbackService.list(db, current.project_id or 0, target_type)
    )


@router.post("/model-evaluations", response_model=R[dict])
def create_model_evaluation(
    payload: ModelEvaluationIn,
    current: CurrentUser = Depends(require_permission("execution:update")),
    db: Session = Depends(get_db),
):
    return R.ok(
        service.PromptEvaluationService.evaluate(
            db, payload.model_dump()
        )
    )


@router.get("/model-evaluations", response_model=R[list])
def list_model_evaluations(
    current: CurrentUser = Depends(require_permission("execution:list")),
    db: Session = Depends(get_db),
):
    return R.ok(service.PromptEvaluationService.list(db))


@router.get("/model-evaluations/regression-check", response_model=R[dict])
def model_eval_regression_check(
    evaluation_suite: str = Query(default=""),
    current: CurrentUser = Depends(require_permission("execution:list")),
    db: Session = Depends(get_db),
):
    return R.ok(
        service.PromptEvaluationService.check_regression(db, evaluation_suite)
    )

File: test-platform-v2/frontend/src/api/aiClosedLoop.ts
import { aitdeV2 } from './missions'

// ── AITDE V3.8 AI QA Closed Loop domain types ──

export interface FailureHypothesis {
  id: number
  run_id: number
  hypothesis_type: string
  classification: string
  summary: string
  confidence: number
  evidence_refs: Record<string, unknown>[]
  suggested_checks: string[]
  model_ref?: string | null
  prompt_version?: string | null
  status: string
  reviewed_by?: number | null
  created_at: string | null
  outcome?: string | null
}

export interface FlakyCluster {
  id: number
  scenario_adapter_id: number
  cluster_key: string
  classification: string
  sample_size: number
  failure_rate: number
  confidence: number
  status: string
}

export interface AiSuggestion {
  id: number
  project_id: number
  mission_id?: number | null
  suggestion_type: string
  target_type: string
  target_id: number
  payload: Record<string, unknown>
  evidence_refs: Record<string, unknown>[]
  confidence: number
  status: string
}

export interface ScenarioGap {
  id: number
  mission_id: number
  gap_type: string
  title: string
  description: string
  source_refs: Record<string, unknown>[]
  evidence_refs: Record<string, unknown>[]
  risk_level: string
  confidence: number
  status: string
}

export interface ModelEvaluation {
  id: number
  evaluation_suite: string
  model_ref: string
  prompt_versions: string[]
  status: string
  metrics: Record<string, unknown>
  artifact_uri?: string | null
  created_at: string | null
}

export interface HumanFeedback {
  id: number
  project_id: number
  mission_id?: number | null
  target_type: string
  target_id: number
  feedback_type: string
  before?: Record<string, unknown> | null
  after?: Record<string, unknown> | null
  reason?: string | null
  created_by: number
}

export type ReviewAction = 'APPROVED' | 'REJECTED'

// ── Failure triage / hypothesis ──

export function triageRun(
  runId: number,
  payload: {
    context?: Record<string, unknown>
    model_ref?: string | null
    prompt_version?: string | null
  },
): Promise<FailureHypothesis> {
  return aitdeV2.post(`/runs/${runId}/triage`, payload)
}

export function listHypotheses(runId: number, signal?: AbortSignal): Promise<FailureHypothesis[]> {
  return aitdeV2.get(`/runs/${runId}/hypotheses`, { signal })
}

export function reviewHypothesis(
  hypothesisId: number,
  payload: { status: string; reviewed_by?: number | null; reason?: string | null },
): Promise<FailureHypothesis> {
  return aitdeV2.post(`/hypotheses/${hypothesisId}/review`, payload)
}

// ── Healing ──

export function applyHealing(
  proposalId: number,
  payload: { approved_by?: number | null; note?: string | null },
): Promise<{
  command_plan_id: number
  command_plan_version_id: number
  version_no: number
  status: string
  old_retained: boolean
}> {
  return aitdeV2.post(`/healing-proposals/${proposalId}/apply`, payload)
}

// ── Flaky / stability ──

export function listFlaky(
  scenarioAdapterId?: number | null,
  signal?: AbortSignal,
): Promise<FlakyCluster[]> {
  const qs = scenarioAdapterId ? `?scenario_adapter_id=${scenarioAdapterId}` : ''
  return aitdeV2.get(`/flaky${qs}`, { signal })
}

export function scenarioStability(scenarioId: number, signal?: AbortSignal): Promise<{ scenario_id: number; clusters: FlakyCluster[] }> {
  return aitdeV2.get(`/scenarios/${scenarioId}/stability`, { signal })
}

// ── Suggestion inbox ──

export function listAiSuggestions(
  status?: string | null,
  signal?: AbortSignal,
): Promise<AiSuggestion[]> {
  const qs = status ? `?status=${status}` : ''
  return aitdeV2.get(`/ai-suggestions${qs}`, { signal })
}

export function reviewAiSuggestion(
  suggestionId: number,
  payload: { status: ReviewAction; reviewed_by?: number | null; reason?: string | null },
): Promise<AiSuggestion> {
  return aitdeV2.post(`/ai-suggestions/${suggestionId}/review`, payload)
}

// ── Scenario gap ──

export function listScenarioGaps(missionId: number, signal?: AbortSignal): Promise<ScenarioGap[]> {
  return aitdeV2.get(`/missions/${missionId}/scenario-gaps`, { signal })
}

export function convertScenarioGap(
  gapId: number,
  payload: { title?: string | null; risk_level?: string; description?: string | null },
): Promise<{ gap_id: number; converted: boolean; title: string; risk_level: string; note: string }> {
  return aitdeV2.post(`/scenario-gaps/${gapId}/convert`, payload)
}

// ── Feedback ──

export function addFeedback(
  payload: {
    project_id?: number
    mission_id?: number | null
    target_type: string
    target_id: number
    feedback_type?: string
    before?: Record<string, unknown> | null
    after?: Record<string, unknown> | null
    reason?: string | null
  },
): Promise<HumanFeedback> {
  return aitdeV2.post('/feedback', payload)
}

export function listFeedback(targetType?: string | null, signal?: AbortSignal): Promise<HumanFeedback[]> {
  const qs = targetType ? `?target_type=${targetType}` : ''
  return aitdeV2.get(`/feedback${qs}`, { signal })
}

// ── Model evaluation ──

export function createModelEvaluation(
  payload: {
    evaluation_suite: string
    model_ref: string
    prompt_versions?: string[]
    metrics?: Record<string, unknown>
    status?: string
    artifact_uri?: string | null
  },
): Promise<ModelEvaluation> {
  return aitdeV2.post('/model-evaluations', payload)
}

export function listModelEvaluations(signal?: AbortSignal): Promise<ModelEvaluation[]> {
  return aitdeV2.get('/model-evaluations', { signal })
}

export function modelEvalRegressionCheck(evaluation_suite: string): Promise<{ ok: boolean; passed: boolean; score: number; threshold: number; reason: string }> {
  return aitdeV2.get(`/model-evaluations/regression-check?evaluation_suite=${encodeURIComponent(evaluation_suite)}`)
}

// ── Label maps ──

export const HYPOTHESIS_STATUS_LABELS: Record<string, { label: string; color: string }> = {
  GENERATED: { label: '自动生成', color: 'bg-status-info-muted text-status-info' },
  REVIEWED: { label: '已复核', color: 'bg-status-warning-muted text-status-warning' },
  CONFIRMED: { label: '已确认', color: 'bg-status-success-muted text-status-success' },
  REJECTED: { label: '已拒绝', color: 'bg-muted text-muted-foreground' },
}

export const SUGGESTION_TYPE_LABELS: Record<string, string> = {
  HEALING: '愈合修复',
  DATA_STRATEGY: '数据策略',
  SCENARIO_GAP: '场景缺口',
  RISK: '风险',
  TRIAGE: '失败归因',
}

export const SUGGESTION_STATUS_LABELS: Record<string, string> = {
  OPEN: '待处理',
  APPROVED: '已批准',
  REJECTED: '已拒绝',
  APPLIED: '已应用',
  EXPIRED: '已过期',
}

export const GAP_TYPE_LABELS: Record<string, string> = {
  PROD_NEW_STATE: '生产新状态',
  HISTORICAL_DEFECT: '历史缺陷',
  REPEATED_BUSINESS_FAIL: '重复业务失败',
  UNCOVERED_CONTRACT_RULE: '未覆盖契约规则',
  UNCOVERED_JOURNEY: '未覆盖旅程',
  NEW_OPENAPI_STATE: '新 OpenAPI 状态',
}
